plugins/legion-go2/fu-legion-go2-device.c

```diff
@@ -0,0 +1,671 @@
+	g_message("start step done: ");
+	
+	// query size step
+	guint maxSize = 0;
+	{
+		g_autoptr(FuStructLegionGo2UpgradeCmd) queryCmd = fu_struct_legion_go2_upgrade_cmd_new();
+		guint8 content[] = { 
+			0x02,
+			FU_LEGION_GO2_UPGRADE_STEP_QUERY_SIZE,
+			0x01
+		};
+		fu_struct_legion_go2_upgrade_cmd_set_report_id(queryCmd, 5);
+		fu_struct_legion_go2_upgrade_cmd_set_length(queryCmd, sizeof(content) + 5);
+		fu_struct_legion_go2_upgrade_cmd_set_device_id(queryCmd, id);
+		fu_struct_legion_go2_upgrade_cmd_set_param(queryCmd, 0x01);
+		fu_struct_legion_go2_upgrade_cmd_set_data(queryCmd, content, sizeof(content), error);
+		if (!fu_legion_go2_device_upgrade_query_size(self, queryCmd->buf, response, error))
+		{
+			return FALSE;
+		}
+		maxSize = response->data[9] << 8 | response->data[10];
+	}
+	fu_progress_step_done(progress);
+	g_message("query size step done: ");
+
+	// write data step
+	{
+		if (maxSize % FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH != 0)
+		{
+			g_message("max size (%u) invalid", maxSize);
+			return FALSE;
+		}
+		guint innerLoopTimes = maxSize / FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH;
+		guint outerLoopTimes = size / maxSize;
+		if ((size % maxSize) != 0)
+		{
+			outerLoopTimes += 1;
+		}
+		guint sendSize = 0;
+		for (guint i = 0; i < outerLoopTimes; ++i)
+		{
+			for (guint j = 0; j < innerLoopTimes; ++j)
+			{
+				if (sendSize >= size)
+				{
+					break;
+				}
+				g_autoptr(FuStructLegionGo2UpgradeCmd) writeCmd = fu_struct_legion_go2_upgrade_cmd_new();
+				guint8 content[FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH + 1] = { 0 };
+				memcpy(content, array->data + sendSize, FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH);
+				content[FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH] = 0x01;
+				fu_struct_legion_go2_upgrade_cmd_set_report_id(writeCmd, 5);
+				fu_struct_legion_go2_upgrade_cmd_set_length(writeCmd, sizeof(content) + 5);
+				fu_struct_legion_go2_upgrade_cmd_set_device_id(writeCmd, id);
+				fu_struct_legion_go2_upgrade_cmd_set_param(writeCmd, 0x02);
+				fu_struct_legion_go2_upgrade_cmd_set_data(writeCmd, content, sizeof(content), error);
+				guint tempSize = sendSize + FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH;
+				gboolean wait = (tempSize % maxSize == 0) || (tempSize >= size);
+				if (!fu_legion_go2_device_upgrade_write_data(self, writeCmd->buf, response, wait, error))
+				{
+					return FALSE;
+				}
+				sendSize += FU_LEGION_GO2_DEVICE_FW_PACKET_LENGTH;
+				if (wait)
+				{
+					guint recievedSize = response->data[10] << 24 | 
+							     response->data[11] << 16 | 
+							     response->data[12] << 8  | 
+							     response->data[13];
+					if (recievedSize != sendSize && recievedSize != size)
+					{
+						g_message("size verify failed(%u %u %u)", sendSize, recievedSize, size);
+						return FALSE;
+					}
+				}
+			}
+		}
+	}
+	fu_progress_step_done(progress);
+	g_message("write data step done: ");
+	
+	// verify step
+	{
+		g_autoptr(FuStructLegionGo2UpgradeCmd) verifyCmd = fu_struct_legion_go2_upgrade_cmd_new();
```

Comment:
this should be called something like `st_verifycmd` -- you need to set up precommit as this would have warned you about most of these.